### ISSUE TITLE
Clarify semantics of calling different poll_* methods from different tasks

### ISSUE THREAD

**Issue by alecmocatta** [rust-lang/futures-rs#2292]
How many `Waker`s should items implementing say `AsyncRead` and `AsyncWrite` have to store? At least some libs<sup>1</sup> store two – a "read waker" and a "write waker". The former is set during `poll_read` and `poll_read_vectored`, and the latter during `poll_write`, `poll_write_vectored`, `poll_flush` and `poll_close`. Is this sufficient?

What about an application that `poll_write`s from one task, but, for example, `poll_flush`es from another, background, task? There's a race between tasks to be the last to set the "write waker", and the losing task may never be able to be woken.

This would be fine, but it's somewhat surprising and it's not totally obvious from the documentation.

---

<sup>1. For example it seems e.g. [`tokio::net::TcpStream`](https://github.com/tokio-rs/tokio/blob/28933599888a88e601acbb11fa824b0ee9f98c6e/tokio/src/io/driver/scheduled_io.rs#L290-L348) and [`tokio-tungstenite`](https://github.com/snapview/tokio-tungstenite/blob/71a5b72123db32b318d48964948fc76c943f1548/src/compat.rs) do this</sup>

**Comment by taiki-e** [comment]
> There's a race between tasks to be the last to set the "write waker", and the losing task may never be able to be woken.

Yeah, it can actually happen, but as far as I know, there is no known AsyncRead/AsyncWrite implementation that actually supports any number of concurrent I/O operations.

See also #2139

**Comment by taiki-e** [comment]
> > There's a race between tasks to be the last to set the "write waker", and the losing task may never be able to be woken.
> 
> Yeah, it can actually happen, but as far as I know, there is no known AsyncRead/AsyncWrite implementation that actually supports any number of concurrent I/O operations.

FYI: tokio has docs that clarify this (https://github.com/tokio-rs/tokio/pull/3310)

> Note that on multiple calls to `poll_peek`, `poll_read` or
> `poll_read_ready`, only the `Waker` from the `Context` passed to the
> most recent call is scheduled to receive a wakeup. (However,
> `poll_write` retains a second, independent waker.)
